docs/manual/scala/guide/cluster/DomainModelling.md

```diff
@@ -1,19 +1,202 @@
+        Effect.reply(replyTo)(Rejected("Quantity must be greater than zero"))
+
+      // an item is deleted by setting it's quantity to 0
+      case UpdateItem(productId, 0, replyTo) if !items.contains(productId) =>
+        Effect.reply(replyTo)(Rejected("Cannot delete item that is not already in cart"))
+
+      case UpdateItem(productId, quantity, replyTo) =>
+        Effect
+          .persist(ItemUpdated(productId, quantity))
+          .thenReply(replyTo) { updatedCart => // updatedCart is the state updated after applying ItemUpdated
+            Accepted
+          }
+
+      // check it out
+      case Checkout(replyTo) =>
+        Effect
+          .persist(CheckedOut)
+          .thenReply(replyTo){ updatedCart => // updated cart is state updated after applying CheckedOut
+            Accepted
+          }
+
+      case Get(replyTo) =>
+        Effect.reply(replyTo)(ShoppingCartSummary(items, status = "open"))
+  }
+}
+
+case class CheckedOutShoppingCart(items: Map[String, Int]) extends ShoppingCart {
+  def applyCommand(cmd: ShoppingCartCommnad) =
+    cmd match {
+      // CheckedOut is a final state, no mutations allowed
+      case UpdateItem(_, _, replyTo) =>
+        Effect.reply(replyTo)(Rejected("Cannot update a checked-out cart"))
+      // CheckedOut is a final state, no mutations allowed
+      case Checkout(replyTo) =>
+        Effect.reply(replyTo)(Rejected("Cannot checkout a checked-out cart"))
+
+      // it is allowed to read it's state though
+      case Get(replyTo) =>
+        Effect
+        .reply(replyTo)(ShoppingCartSummary(items, status = "checked-out"))
+    }
+
+}
+```
+
+Command handlers are the meat of the model. They encode the business rules of your model and act as a guardian of the model consistency. The command handler must first validate that the incoming command can be applied to the current model state. In case of successful validation, one or more event expressing the mutations are persisted. Once the events are persisted, they are applied to the state producing a new valid state.
+
+Because an Aggregate is intended to model a consistency boundary, it's not recommended to validate commands using data that's not available in scope. Any decision should be solely based on the data passed in the commands and the state of the aggregate. Any external call should be considered a smell because it means that the aggregate is not in full control of the invariants it's supposed to be protecting.
+
+There two ways of sending back a reply: using `Effect.reply` and `Effect.persist(...).thenReply`. The first one is available directly on `Effect` and should be used when you reply without persisting any event. In this case, you can use the available state in scope because it's guaranteed to not have changed. The second variant should be used when you have persisted one or more events. The updated state is then made available to you on the function used to define the reply.
```

Comment:
Oh, gods, there's an error in the suggestion I made earlier:

```suggestion
There are two ways of sending back a reply: using `Effect.reply` and `Effect.persist(...).thenReply`. The first one is available directly on `Effect` and should be used when you reply without persisting any event. In this case, you can use the available state in scope because it's guaranteed to not have changed. The second variant should be used when you have persisted one or more events. The updated state is then made available to you on the function used to define the reply.
```